compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotSerializer.kt

```diff
@@ -252,3 +258,46 @@ object InaccessibleClassSnapshotExternalizer : DataExternalizer<InaccessibleClas
         return InaccessibleClassSnapshot
     }
 }
+
+private object ClassIdExternalizer : DataExternalizer<ClassId> {
+
+    override fun save(output: DataOutput, classId: ClassId) {
+        FqNameExternalizer.save(output, classId.packageFqName)
+        FqNameExternalizer.save(output, classId.relativeClassName)
+        output.writeBoolean(classId.isLocal)
+    }
+
+    override fun read(input: DataInput): ClassId {
+        return ClassId(
+            // To reduce memory usage, apply object interning to package names as classes can share the same package name
+            /* packageFqName */ fqNameInterner.intern(FqNameExternalizer.read(input)),
+            /* relativeClassName */ FqNameExternalizer.read(input),
+            /* isLocal */ input.readBoolean()
+        )
+    }
+}
+
+private object FqNameExternalizer : DataExternalizer<FqName> {
+
+    override fun save(output: DataOutput, fqName: FqName) {
+        output.writeString(fqName.asString())
+    }
+
+    override fun read(input: DataInput): FqName {
+        return FqName(input.readString())
+    }
+}
+
+private object JvmClassNameExternalizer : DataExternalizer<JvmClassName> {
+
+    override fun save(output: DataOutput, jvmClassName: JvmClassName) {
+        output.writeString(jvmClassName.internalName)
+    }
+
+    override fun read(input: DataInput): JvmClassName {
```

Comment:
We want to intern as soon as we read data, right? I think you can accomplish that by interning directly in FqNameExternalizer.read() and JvmClassNameExternalizer.read(). Also, this allows you to keep the intern pools private to those classes.

Reply:
The issue with that approach is that unique objects would also be interned which means we would intern more stuff than necessary, wasting both CPU + memory.

For example, `FqName`s when used as package names are usually shared, but when used as simple class names are usually not shared. `JvmClassName`s when used as supertypes are usually shared, but when used as class IDs are usually not shared.

I don't have a strong preference though, so let me know if you think keeping the intern pools in the externalizers is still better.

Reply:
Interning only for commonly shared data makes sense, so let's continue doing that.

An improvement could be introducing `JvmClassNameExternalizerWithInterning` which takes care of interning (the same for FqName), so that interning logic does not leak.